## Dropout Regularization In Deep Neural Network

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv("sonar_dataset.csv", header=None)
df

0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      51      52      53      54      55      56      57  \
0    0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1    0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2    0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3    0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4    0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

         58      59  60  
0    0.0090  0.0032   R  
1    0.0052  0.0044   R  
2    0.0095  0.0078   R  
3    0.0040  0.0117   R  
4    0.0107  0.0094   R  
..      ...     ...  ..  
203  0.0193  0.0157   M  
204  0.0062  0.0067   M  
205  0.0077  0.0031   M  
206  0.0036  0.0048   M  
207  0.0061  0.0115   M  

[208 rows x 61 columns]

In [7]:
df.shape

(208, 61)

In [8]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [9]:

df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [10]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

In [11]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [12]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
32   1
172  0
67   1
100  0
58   1

In [13]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [14]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [15]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [16]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

## Using Deep Learning Model

### Model without Dropout Layer

In [17]:
import tensorflow as tf
from tensorflow import keras

In [18]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'), 
     keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100)

Epoch 1/100
20/20 [==============================] - 1s 3ms/step - loss: 0.6877 - accuracy: 0.5513
Epoch 2/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6678 - accuracy: 0.6090
Epoch 3/100
20/20 [==============================] - 0s 11ms/step - loss: 0.6478 - accuracy: 0.6410
Epoch 4/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6261 - accuracy: 0.6795
Epoch 5/100
20/20 [==============================] - 0s 5ms/step - loss: 0.5822 - accuracy: 0.6923
Epoch 6/100
20/20 [==============================] - 0s 4ms/step - loss: 0.5273 - accuracy: 0.8077
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4739 - accuracy: 0.8141
Epoch 8/100
20/20 [==============================] - 0s 5ms/step - loss: 0.4369 - accuracy: 0.8205
Epoch 9/100
20/20 [==============================] - 0s 4ms/step - loss: 0.4244 - accuracy: 0.8526
Epoch 10/100
20/20 [==============================] - 0s 5ms/step - loss: 0.3807 - accuracy: 0.8782
Epoch 11

20/20 [==============================] - 0s 4ms/step - loss: 0.0038 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0038 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 3ms/step - loss: 0.0033 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0032 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0031 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 5ms/step - loss: 0.0030 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0030 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0030 - accuracy: 1.0000
Epoch 91/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0028 - accuracy: 1.0000
Epoch 92/100
20/20 [==============================] - 0s 2ms/step - loss: 0.0026 - accuracy: 1.0000
Epoch 93/100


In [16]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 1s 5ms/step - loss: 0.9907 - accuracy: 0.7308


[0.9906792044639587, 0.7307692170143127]

Training Accuracy >>> Test Accuracy

In [17]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 5ms/step
[1.2403591e-07 9.7085506e-01 8.4299350e-01 2.0441705e-04 9.9999934e-01
 9.6807325e-01 1.9068021e-01 1.0000000e+00 1.8009076e-05 9.9999976e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [18]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [19]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.70      0.85      0.77        27
           1       0.79      0.60      0.68        25

    accuracy                           0.73        52
   macro avg       0.74      0.73      0.72        52
weighted avg       0.74      0.73      0.73        52



### Model with Dropout Layer

In [21]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100)

Epoch 1/100
20/20 [==============================] - 1s 5ms/step - loss: 0.7129 - accuracy: 0.5000
Epoch 2/100
20/20 [==============================] - 0s 4ms/step - loss: 0.7025 - accuracy: 0.4167
Epoch 3/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6924 - accuracy: 0.4936
Epoch 4/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6753 - accuracy: 0.5577
Epoch 5/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6892 - accuracy: 0.5256
Epoch 6/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6691 - accuracy: 0.5449
Epoch 7/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6575 - accuracy: 0.6090
Epoch 8/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6747 - accuracy: 0.5833
Epoch 9/100
20/20 [==============================] - 0s 10ms/step - loss: 0.6358 - accuracy: 0.6154
Epoch 10/100
20/20 [==============================] - 0s 7ms/step - loss: 0.6423 - accuracy: 0.6410
Epoch 11

20/20 [==============================] - 0s 5ms/step - loss: 0.1304 - accuracy: 0.9615
Epoch 84/100
20/20 [==============================] - 0s 4ms/step - loss: 0.1541 - accuracy: 0.9487
Epoch 85/100
20/20 [==============================] - 0s 5ms/step - loss: 0.1258 - accuracy: 0.9551
Epoch 86/100
20/20 [==============================] - 0s 5ms/step - loss: 0.1897 - accuracy: 0.9551
Epoch 87/100
20/20 [==============================] - 0s 4ms/step - loss: 0.1304 - accuracy: 0.9423
Epoch 88/100
20/20 [==============================] - 0s 4ms/step - loss: 0.1395 - accuracy: 0.9551
Epoch 89/100
20/20 [==============================] - 0s 3ms/step - loss: 0.1692 - accuracy: 0.9423
Epoch 90/100
20/20 [==============================] - 0s 3ms/step - loss: 0.1080 - accuracy: 0.9744
Epoch 91/100
20/20 [==============================] - 0s 4ms/step - loss: 0.1616 - accuracy: 0.9359
Epoch 92/100
20/20 [==============================] - 0s 3ms/step - loss: 0.1685 - accuracy: 0.9359
Epoch 93/100


In [22]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 2ms/step - loss: 0.7104 - accuracy: 0.7692


[0.7104119658470154, 0.7692307829856873]

Training Accuracy is still good but Test Accuracy Improved

In [22]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 3ms/step
[8.6528411e-09 9.5997739e-01 9.8371750e-01 2.0746212e-04 9.9988854e-01
 9.9099231e-01 2.4145672e-01 9.9996561e-01 8.5609790e-05 9.9995142e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [23]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.73      0.89      0.80        27
           1       0.84      0.64      0.73        25

    accuracy                           0.77        52
   macro avg       0.78      0.76      0.76        52
weighted avg       0.78      0.77      0.77        52



You can see that by using dropout layer test accuracy increased from 0.77 to 0.79